File: src/mcp/agentos.ts
/**
 * AgentOS MCP — the control-plane tool set agents get (§4 "Built-in MCPs").
 * Read/write the current task, mark status (approval gates enforced), spawn
 * collaborators (collaboration list enforced), read goal context, append to
 * the goal progress log, read project metadata.
 */
import { z } from "zod";
import { eq, and } from "drizzle-orm";
import { agents, projects, goals as goalsTable, files } from "../db/schema.js";
import { createMcpServer, type McpRuntime } from "./context.js";
import type { McpServer } from "./context.js";
import { checkFsOp } from "../acl/grants.js";
import type { TaskStatus } from "../domain/types.js";

export function createAgentosMcpServer(): McpServer {
  return createMcpServer("agentos", [
    {
      name: "tasks.current",
      description: "Read the current task (or null when this session has no task).",
      inputSchema: z.object({}),
      handler: async (rt) => {
        if (!rt.manifest.task) return { task: null };
        const task = await rt.services.tasks.get(rt.manifest.task.id);
        if (!task) return { task: null };
        const attachments = await Promise.all(
          task.attachmentIds.map((id) => rt.services.files.getById(id)),
        );
        return {
          task: {
            id: task.id,
            name: task.name,
            description: task.description,
            status: task.status,
            approvalGate: task.approvalGate,
            activity: task.activity,
            attachments: attachments.filter(Boolean),
          },
        };
      },
    },
    {
      name: "tasks.append_activity",
      description:
        "Append a progress note to the current task's activity log. Use this as your running work plan: write the ordered steps and what 'done' means at the start, then keep it current and check items off as you go.",
      inputSchema: z.object({ message: z.string().min(1) }),
      handler: async (rt, { message }) => {
        const taskId = rt.manifest.task?.id;
        if (!taskId) throw new Error("this session has no task");
        const task = await rt.services.tasks.appendActivity(
          taskId,
          rt.manifest.agent.name,
          message,
        );
        return { ok: true, activity: task.activity };
      },
    },
    {
      name: "tasks.set_status",
      description:
        "Move the current task between todo/doing/review/done. Lifecycle: set 'doing' when you start, 'review' when your deliverable is ready for a human, 'done' only after your exit criteria are verified. 'done' is REFUSED when the task has an approval gate — only the human can mark those done; leave them in 'review' and inbox the human.",
      inputSchema: z.object({ status: z.enum(["todo", "doing", "review", "done"]) }),
      handler: async (rt, { status }) => {
        const taskId = rt.manifest.task?.id;
        if (!taskId) throw new Error("this session has no task");
        const task = await rt.services.tasks.setStatus(taskId, status as TaskStatus, {
          actor: "agent",
          agentId: rt.manifest.agent.id,
          sessionId: rt.sessionId,
        });
        return { ok: true, status: task.status, approvalGate: task.approvalGate };
      },
    },
    {
      name: "tasks.attach",
      description: "Attach a file (by virtual path) to the current task.",
      inputSchema: z.object({ filePath: z.string() }),
      handler: async (rt, { filePath }) => {
        const taskId = rt.manifest.task?.id;
        if (!taskId) throw new Error("this session has no task");
        // Attaching is a read of the file: enforce the agent's filesystem
        // ACL the same way fs.read does (§5.7 — least privilege everywhere).
        const fsOp = checkFsOp(rt.manifest.filesystemGrants, "read", filePath);
        if (!fsOp.ok) throw new Error(`attachment denied: ${fsOp.reason}`);
        const f = (
          await rt.services.db
            .select()
            .from(files)
            .where(and(eq(files.projectId, rt.manifest.projectId), eq(files.path, filePath)))
            .get()
        );
        if (!f) throw new Error(`file not found: ${filePath}`);
        const task = await rt.services.tasks.attachFile(taskId, f.id);
        return { ok: true, attachmentIds: task.attachmentIds };
      },
    },
    {
      name: "collaborators.spawn",
      description:
        "Spawn a collaborator subtask. Only agents on this agent's collaboration list may be spawned (§5.10). Give a tight brief: the artifact to produce, the deliverable you expect, and the exit criteria you will check — the collaborator has no other context.",
      inputSchema: z.object({ agentName: z.string(), brief: z.string().min(1) }),
      handler: async (rt, { agentName, brief }) => {
        const list = rt.manifest.collaborationList;
        if (!list.includes(agentName)) {
          throw new Error(
            `collaborator "${agentName}" is not on the collaboration list [${list.join(", ")}]`,
          );
        }
        const agent = await rt.services.db
          .select()
          .from(agents)
          .where(and(eq(agents.projectId, rt.manifest.projectId), eq(agents.name, agentName)))
          .get();
        if (!agent) throw new Error(`unknown agent "${agentName}"`);
        const task = await rt.services.tasks.spawnCollaborator({
          projectId: rt.manifest.projectId,
          agentId: agent.id,
          agentName,
          brief,
          parentTaskId: rt.manifest.task?.id ?? null,
          parentSessionId: rt.sessionId,
        });
        return { ok: true, taskId: task.id, name: task.name };
      },
    },
    {
      name: "goals.current",
      description: "Read the goal context (title, spec, definition of done) for this session.",
      inputSchema: z.object({}),
      handler: async (rt) => {
        if (!rt.manifest.goal) return { goal: null };
        const g = await rt.services.goals.get(rt.manifest.goal.id);
        if (!g) return { goal: null };
        return {
          goal: {
            id: g.id,
            title: g.title,
            spec: g.spec,
            definitionOfDone: g.definitionOfDone,
            progressLog: g.progressLog,
          },
        };
      },
    },
    {
      name: "goals.append_progress",
      description:
        "Append an entry to the goal's append-only progress log (a human-readable trail of what happened). Marking a definition-of-done item done is separate — use goals.complete_dod_item once the item is actually verified.",
      inputSchema: z.object({ entry: z.string().min(1) }),
      handler: async (rt, { entry }) => {
        if (!rt.manifest.goal) throw new Error("this session has no goal");
        const g = await rt.services.goals.appendProgress(rt.manifest.goal.id, entry);
        return { ok: true, progressLog: g.progressLog };
      },
    },
    {
      name: "goals.complete_dod_item",
      description:
        "Mark ONE definition-of-done item as satisfied after you have actually finished AND verified it. This is the only way an item becomes done — the orchestrator never infers completion from progress-log text. Pass the item text exactly as listed. The goal completes only when every item is marked this way.",
      inputSchema: z.object({ item: z.string().min(1) }),
      handler: async (rt, { item }) => {
        if (!rt.manifest.goal) throw new Error("this session has no goal");
        const g = await rt.services.goals.completeDoDItem(rt.manifest.goal.id, item);
        return { ok: true, definitionOfDone: g.definitionOfDone };
      },
    },
    {
      name: "projects.meta",
      description: "Read project metadata (name, slug). Never contains secrets.",
      inputSchema: z.object({}),
      handler: async (rt) => {
        const p = await rt.services.db
          .select()
          .from(projects)
          .where(eq(projects.id, rt.manifest.projectId))
          .get();
        if (!p) throw new Error("project not found");
        return { project: { id: p.id, name: p.name, slug: p.slug } };
      },
    },
  ]);
}

File: agentos/mcp/agentos.py
from __future__ import annotations
"""AgentOS MCP server. Port of src/mcp/agentos.ts."""
from agentos.mcp.context import McpServer, McpTool, create_mcp_server
from agentos.acl.grants import check_fs_op


def create_agentos_mcp_server() -> McpServer:
    async def tasks_current(rt, args):
        if not rt.manifest.task:
            return {"task": None}
        task = await rt.services.tasks.get(rt.manifest.task.id)
        if not task:
            return {"task": None}
        attachments = [
            await rt.services.files.get_by_id(fid) for fid in task.get("attachmentIds", [])
        ]
        return {"task": {
            "id": task["id"], "name": task["name"], "description": task["description"],
            "status": task["status"], "approvalGate": task["approvalGate"],
            "activity": task["activity"], "attachments": [a for a in attachments if a],
        }}

    async def tasks_append_activity(rt, args):
        task_id = rt.manifest.task.id if rt.manifest.task else None
        if not task_id:
            raise ValueError("this session has no task")
        task = await rt.services.tasks.append_activity(task_id, rt.manifest.agent.name, args["message"])
        return {"ok": True, "activity": task["activity"]}

    async def tasks_set_status(rt, args):
        task_id = rt.manifest.task.id if rt.manifest.task else None
        if not task_id:
            raise ValueError("this session has no task")
        task = await rt.services.tasks.set_status(
            task_id, args["status"],
            actor="agent", agent_id=rt.manifest.agent.id, session_id=rt.session_id,
        )
        return {"ok": True, "status": task["status"], "approvalGate": task["approvalGate"]}

    async def tasks_attach(rt, args):
        task_id = rt.manifest.task.id if rt.manifest.task else None
        if not task_id:
            raise ValueError("this session has no task")
        ok, reason = check_fs_op(rt.manifest.filesystemGrants, "read", args["filePath"])
        if not ok:
            raise PermissionError(f"attachment denied: {reason}")
        from agentos.db.models import File as FileRow
        from agentos.db.client import get_session as _get_session
        from sqlalchemy import select, and_
        async with _get_session() as db:
            result = await db.execute(
                select(FileRow).where(
                    and_(FileRow.project_id == rt.manifest.projectId, FileRow.path == args["filePath"])
                )
            )
            f = result.scalar_one_or_none()
        if not f:
            raise FileNotFoundError(f"file not found: {args['filePath']}")
        from agentos.db.models import Task as TaskRow
        async with _get_session() as db:
            result = await db.execute(select(TaskRow).where(TaskRow.id == task_id))
            task_row = result.scalar_one_or_none()
            if task_row:
                ids = list(task_row.attachment_ids or [])
                if f.id not in ids:
                    ids.append(f.id)
                    task_row.attachment_ids = ids
                    await db.commit()
        return {"ok": True}

    async def collaborators_spawn(rt, args):
        list_ = rt.manifest.collaborationList
        if args["agentName"] not in list_:
            raise PermissionError(
                f'collaborator "{args["agentName"]}" not on collaboration list [{", ".join(list_)}]'
            )
        task = await rt.spawn_collaborator(args["agentName"], args["brief"])
        return {"ok": True, "taskId": task["id"], "name": task["name"]}

    async def goals_current(rt, args):
        if not rt.manifest.goal:
            return {"goal": None}
        g = await rt.services.goals.get(rt.manifest.goal.id)
        if not g:
            return {"goal": None}
        return {"goal": {"id": g["id"], "title": g["title"], "spec": g["spec"],
                          "definitionOfDone": g["definitionOfDone"],
                          "progressLog": g["progressLog"]}}

    async def goals_append_progress(rt, args):
        if not rt.manifest.goal:
            raise ValueError("this session has no goal")
        g = await rt.services.goals.append_progress(rt.manifest.goal.id, args["entry"])
        return {"ok": True, "progressLog": g["progressLog"]}

    async def goals_complete_dod_item(rt, args):
        if not rt.manifest.goal:
            raise ValueError("this session has no goal")
        g = await rt.services.goals.complete_dod_item(rt.manifest.goal.id, args["item"])
        return {"ok": True, "definitionOfDone": g["definitionOfDone"]}

    async def projects_meta(rt, args):
        from agentos.db.models import Project as ProjectRow
        from agentos.db.client import get_session as _get_session
        from sqlalchemy import select
        async with _get_session() as db:
            result = await db.execute(select(ProjectRow).where(ProjectRow.id == rt.manifest.projectId))
            p = result.scalar_one_or_none()
        if not p:
            raise ValueError("project not found")
        return {"project": {"id": p.id, "name": p.name, "slug": p.slug}}

    return create_mcp_server("agentos", [
        McpTool("tasks.current", "Read the current task.", tasks_current),
        McpTool("tasks.append_activity", "Append a progress note to the current task's activity log.", tasks_append_activity),
        McpTool("tasks.set_status", "Move the current task status (todo/doing/review/done). Gated tasks cannot be set to done by agents.", tasks_set_status),
        McpTool("tasks.attach", "Attach a file by virtual path to the current task.", tasks_attach),
        McpTool("collaborators.spawn", "Spawn a collaborator subtask (collaboration list enforced).", collaborators_spawn),
        McpTool("goals.current", "Read the goal context for this session.", goals_current),
        McpTool("goals.append_progress", "Append to the goal's progress log.", goals_append_progress),
        McpTool("goals.complete_dod_item", "Mark one DoD item done after verifying it.", goals_complete_dod_item),
        McpTool("projects.meta", "Read project metadata (name, slug).", projects_meta),
    ])
